packages/clay-data-provider/src/useResource.ts

```diff
@@ -0,0 +1,228 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import Uri from 'metal-uri';
+import {
+	FetchPolicy,
+	IDataProvider,
+	NetworkStatus,
+	SYMBOL_ORIGIN,
+	TVariables,
+} from './types';
+import {timeout} from './util';
+import {useCache} from './useCache';
+import {useDebounce} from '@clayui/shared';
+import {useEffect, useRef, useState} from 'react';
+
+interface IResource extends IDataProvider {
+	onNetworkStatusChange?: (status: NetworkStatus) => void;
+}
+
+const useResource = ({
+	fetchDelay = 300,
+	fetchOptions,
+	fetchPolicy = FetchPolicy.NoCache,
+	fetchTimeout = 6000,
+	link,
+	onNetworkStatusChange = () => {},
+	pollInterval = 0,
+	storage = {
+		/**
+		 * This will ensure that we know that the storage
+		 * reference is not external, otherwise the cache
+		 * will persist by the application.
+		 */
+		[SYMBOL_ORIGIN]: true,
+	},
+	fetchRetry = {},
+	storageMaxSize = 20,
+	variables = null,
+}: IResource) => {
+	const [resource, setResource] = useState<any>(null);
+
+	let networkStatus = useRef<NetworkStatus>(NetworkStatus.Unused).current;
+
+	let pollingIntervalId = useRef<null | NodeJS.Timeout>(null).current;
+
+	let retryDelayIntervalId = useRef<null | NodeJS.Timeout>(null).current;
+
+	// A flag to identify if the first rendering happened to avoid
+	// two requests.
+	const firstRenderRef = useRef<boolean>(true);
+
+	const cache = useCache(
+		fetchPolicy,
+		storage,
+		storageMaxSize,
+		link,
+		variables
+	);
+
+	const debouncedVariablesChange = useDebounce(variables, fetchDelay);
+
+	const setNetworkStatus = (status: NetworkStatus) => {
+		networkStatus = status;
+		onNetworkStatusChange(status);
+	};
+
+	const handleRefetch = () => {
+		doFetch();
+		setNetworkStatus(NetworkStatus.Refetch);
+	};
+
+	const cleanRetry = () => {
+		if (retryDelayIntervalId) {
+			clearInterval(retryDelayIntervalId);
+		}
+	};
+
+	const getRetryDelay = (retryAttempts: number) => {
+		const {delay = {}} = fetchRetry;
+		const {initial = 300, jitter = true} = delay;
+
+		const baseDelay = jitter ? initial : initial / 2;
+
+		let value = baseDelay * 2 ** retryAttempts;
+
+		if (jitter) {
+			value = Math.random() * value;
+		}
+
+		return value;
+	};
+
+	const handleFetchRetry = (err: any, retryAttempts: number) => {
+		const {attempts = 5} = fetchRetry;
+
+		cleanRetry();
+
+		if (attempts > 0 && retryAttempts < attempts) {
+			const delay = getRetryDelay(retryAttempts);
+
+			// eslint-disable-next-line no-console
+			console.warn(
```

Comment:
Can we use the `warning` package here since we have been using that in other packags?

Reply:
For this case I think not, because I just want to give a warning and only then I will throw an error if this persists. The `warning` package will throw an error.